### PR TITLE
WIP, [x86/linux] Apply call hoist

### PR DESCRIPTION
**Pull request by YongseopKim**
While FOA is implemented(https://github.com/dotnet/coreclr/pull/10034), try to apply call hoist on x86/linux

### CODE REVIEW

**Comment by YongseopKim** [comment]
cc @seanshpark @parjong

**Comment by YongseopKim** [comment]
current status
- can reach codegen phase
- but, below assert failure has produced
```
Assert failure(PID 305 [0x00000131], Thread: 305 [0x0131]): Assertion failed 'genStackLevel == 0' in 'ConsoleApplication.Program:CallDoIt()' (IL size 20)

    File: /home/dragon/nfs/coreclr/src/jit/codegenlinear.cpp Line: 512
    Image: /home/dragon/nfs/issues/call-hoist/coreoveray_x86_callhoist_nonfoa/corerun
```

jit-dump of morph phase
```
fgMorphTree BB02, stmt 5 (before)
               [000021] --CXG-------             *  call      void   ConsoleApplication.Program.DoIt
               [000009] ------------ arg0        +--*  lclVar    int    V00 loc0
               [000013] ---XG------- arg1        \--*  obj(20)   struct
               [000011] ------------                |     /--*  const     int    4 Fseq[#FirstElem]
               [000012] ----G-------                |  /--*  +         byref
               [000010] ----G-------                |  |  \--*  field     ref    Zero
               [000020] --CXG-------                \--*  comma     byref
               [000019] H-CXG-------                   \--*  call help int    HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
               [000015] ------------ arg0                 +--*  const     int    0xF5A460F0
               [000016] ------------ arg1                 \--*  const     int    1

Sorting the arguments:
Deferred argument ('ecx'):
               [000015] -----+------             *  const     int    0xF5A460F0
Replaced with placeholder node:
               [000038] ----------L-             *  argPlace  int
Deferred argument ('edx'):
               [000016] -----+------             *  const     int    1
Replaced with placeholder node:
               [000040] ----------L-             *  argPlace  int

Shuffled argument table:    ecx edx
fgArgTabEntry[arg 0, ecx, regs=1, align=1, lateArgInx=0, processed]
fgArgTabEntry[arg 1, edx, regs=1, align=1, lateArgInx=1, processed]
Upping fgPtrArgCntMax from 0 to 5

Sorting the arguments:
Argument with 'side effect'...
               [000013] --CXG+------             *  obj(20)   struct
               [000011] -----+------             |     /--*  const     int    4 Fseq[#FirstElem]
               [000012] ----G+------             |  /--*  +         byref
               [000010] ----G+------             |  |  \--*  clsVar    ref    Hnd=0xf5a46b88 Fseq[Zero]
               [000020] --CXG+------             \--*  comma     byref
               [000019] H-CXG+------                \--*  call help int    HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
               [000015] -----+------ arg0 in ecx       +--*  const     int    0xF5A460F0
               [000016] -----+------ arg1 in edx       \--*  const     int    1

lvaGrabTemp returning 1 (V01 tmp0) called for argument with side effect.

fgMorphCopyBlock:block assignment to morph:
               [000013] --CXG+-N----             /--*  obj(20)   struct
               [000011] -----+------             |  |     /--*  const     int    4 Fseq[#FirstElem]
               [000012] ----G+------             |  |  /--*  +         byref
               [000010] ----G+------             |  |  |  \--*  clsVar    ref    Hnd=0xf5a46b88 Fseq[Zero]
               [000020] --CXG+------             |  \--*  comma     byref
               [000019] H-CXG+------             |     \--*  call help int    HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
               [000015] -----+------ arg0 in ecx |        +--*  const     int    0xF5A460F0
               [000016] -----+------ arg1 in edx |        \--*  const     int    1
               [000044] -ACXG---R---             *  =         struct (copy)
               [000042] D------N----             \--*  lclVar    struct V01 tmp0
 with no promoted structs this requires a CopyBlock.

Local V01 should not be enregistered because: written in a block op

 Evaluate to a temp:
               [000013] --CXG+-N----             /--*  indir     struct
               [000011] -----+------             |  |     /--*  const     int    4 Fseq[#FirstElem]
               [000012] ----G+------             |  |  /--*  +         byref
               [000010] ----G+------             |  |  |  \--*  clsVar    ref    Hnd=0xf5a46b88 Fseq[Zero]
               [000020] --CXG+------             |  \--*  comma     byref
               [000019] H-CXG+------             |     \--*  call help int    HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
               [000015] -----+------ arg0 in ecx |        +--*  const     int    0xF5A460F0
               [000016] -----+------ arg1 in edx |        \--*  const     int    1
               [000044] -ACXG---R-L-             *  =         struct (copy)
               [000042] D------N----             \--*  lclVar    struct V01 tmp0
Deferred argument ('ecx'):
               [000009] -----+------             *  lclVar    int    V00 loc0
Replaced with placeholder node:
               [000047] ----------L-             *  argPlace  int

Shuffled argument table:    ecx
fgArgTabEntry[arg 1, numSlots=5, slotNum=0, align=1, lateArgInx=0, tmpNum=V01, isTmp, processed]
fgArgTabEntry[arg 0, ecx, regs=1, align=1, lateArgInx=1, processed]

fgMorphTree BB02, stmt 5 (after)
               [000021] --CXG+------             *  call      void   ConsoleApplication.Program.DoIt
               [000013] --CXG+-N----             |  /--*  indir     struct
               [000011] -----+------             |  |  |     /--*  const     int    4 Fseq[#FirstElem]
               [000012] ----G+------             |  |  |  /--*  +         byref
               [000010] ----G+------             |  |  |  |  \--*  clsVar    ref    Hnd=0xf5a46b88 Fseq[Zero]
               [000020] --CXG+------             |  |  \--*  comma     byref
               [000019] H-CXG+------             |  |     \--*  call help int    HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
               [000015] -----+------ arg0 in ecx |  |        +--*  const     int    0xF5A460F0
               [000016] -----+------ arg1 in edx |  |        \--*  const     int    1
               [000044] -ACXG---R-L- arg1 SETUP  +--*  =         struct (copy)
               [000042] D------N----             |  \--*  lclVar    struct V01 tmp0
               [000050] ------------ arg1 in STK +--*  obj(20)   struct
               [000049] L-----------             |  \--*  addr      int
               [000045] ------------             |     \--*  lclVar    struct V01 tmp0
               [000009] -----+------ arg0 in ecx \--*  lclVar    int    V00 loc0
```

**Comment by YongseopKim** [comment]
Current status
- The simple application is executed normally.
  ```cs
  using System;
  
  namespace ConsoleApplication
  {
	public struct MyStruct
	{
		public static readonly MyStruct Zero = new MyStruct();
		internal int _tick1;
		internal int _tick2;
		internal int _tick3;
		internal int _tick4;
		internal int _tick5;
	}

	public class Program
	{
		public static void DoIt(int a, MyStruct b) {}
		public static void CallDoIt()
		{
			int intValue = 0x1234567;
			DoIt(intValue, MyStruct.Zero);
		}
		public static void Main()
		{
			CallDoIt();
		}
	}
  }
  ```
- but, JIT/Methodical/fp/exgen/200w1d-02_cs_do/200w1d-02_cs_do.exe is not executed.
  ```
  Assert failure(PID 320 [0x00000140], Thread: 320 [0x0140]): Assertion failed 'lateList != nullptr && lateNode != nullptr' in 'testout1:Func_0():int' (IL size 2975)

    File: /home/dragon/nfs/coreclr/src/jit/morph.cpp Line: 4488
    Image: /home/dragon/nfs/issues/call-hoist/coreoveray_x86_callhoist_nonfoa/corerun
  ```

File: src/jit/codegenxarch.cpp [hunk 1]
```diff
@@ -4896,8 +4896,34 @@ void CodeGen::genCallInstruction(GenTreeCall* call)
                 stackArgBytes += genTypeSize(genActualType(arg->TypeGet()));
             }
         }
+
```
Comment (parjong) [inline src/jit/codegenxarch.cpp]: It would be better not to introduce irrelevant formatting changes here.

**Comment by karelz** [comment]
What is status of this PR? What is it blocked on? Code review?

**Comment by YongseopKim** [comment]
@karelz the status of thie PR is stopped because of implementing problem. I'll reopen this after solving it. Thank you for your mention. ;)

**Pull request closed without merging**